src/libraries/Microsoft.PowerFx.Core/Public/Config/ReadOnlySymbolTable.cs

```diff
@@ -0,0 +1,300 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.PowerFx.Core;
+using Microsoft.PowerFx.Core.App;
+using Microsoft.PowerFx.Core.Binding;
+using Microsoft.PowerFx.Core.Binding.BindInfo;
+using Microsoft.PowerFx.Core.Entities;
+using Microsoft.PowerFx.Core.Functions;
+using Microsoft.PowerFx.Core.Types;
+using Microsoft.PowerFx.Core.Types.Enums;
+using Microsoft.PowerFx.Core.Utils;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx
+{
+    /// <summary>
+    /// ReadOnly version of a Symbol Table. This feeds functions, variables, enums, etc into
+    /// the binder.   
+    /// See <see cref="SymbolTable"/> for mutable version. 
+    /// </summary>
+    [DebuggerDisplay("{_debugName}")]
+    [ThreadSafeImmutable]
+    public class ReadOnlySymbolTable : INameResolver, IGlobalSymbolNameResolver
+    {
+        // Changed on each update. 
+        // Host can use to ensure that a symbol table wasn't mutated on us.                 
+        private protected VersionHash _version = VersionHash.New();
+
+        /// <summary>
+        /// This can be compared to determine if the symbol table was mutated during an operation. 
+        /// </summary>
+        internal virtual VersionHash VersionHash => _parent == null ?
+            _version : _version.Combine(_parent.VersionHash);
+
+        /// <summary>
+        /// Notify the symbol table has changed. 
+        /// </summary>
+        public void Inc()
```

Comment:
Why is this public?

Reply:
it lets callers invalidate a symbol table. 
This can be useful if the symbol table refers to some external thing and that has changed.